odl/contrib/solvers/spdhg/stochastic_primal_dual_hybrid_gradient.py

```diff
@@ -0,0 +1,600 @@
+    A : functions
+        Operators A[i] : X -> Y[i] that posses adjoints: A[i].adjoint
+    tau : scalar
+        Initial step size for primal variable.
+    sigma_tilde : scalar
+        Related to initial step size for dual variable.
+    niter : int
+        Number of iterations
+    extra: list
+        List of local extrapolation paramters for every index i.
+    prob: list
+        List of probabilities that an index i is selected each iteration.
+    mu: list
+        List of strong convexity constants of f*, i.e. mu[i] is the strong
+        convexity constant of f*[i].
+    fun_select : function
+        Function that selects blocks at every iteration IN -> {1,...,n}.
+
+    Other Parameters
+    ----------------
+    y: dual variable
+        Dual variable is part of a product space
+    z: variable
+        Adjoint of dual variable, z = A^* y.
+    callback : callable, optional
+        Function called with the current iterate after each iteration.
+
+    References
+    ----------
+    [CERS2017] Chambolle, A., Ehrhardt, M. J., Richtárik, P., and
+    Schoenlieb, C.-B. *Stochastic Primal-Dual Hybrid Gradient Algorithm with
+    Arbitrary Sampling and Imaging Applications*.
+    ArXiv: http://arxiv.org/abs/1706.04957 (2017).
+    """
+
+    # Callback object
+    callback = kwargs.pop('callback', None)
+    if callback is not None and not callable(callback):
+        raise TypeError('`callback` {} is not callable'
+                        ''.format(callback))
+
+    # Dual variable
+    y = kwargs.pop('y', None)
+    if y is None:
+        y = A.range.zero()
+
+    # Adjoint of dual variable
+    z = kwargs.pop('z', None)
+    if z is None and y.norm() == 0:
+        z = A.domain.zero()
+
+    # Initialize variables
+    z_relax = z.copy()
+    dz = A.domain.element()
+    y_old = A.range.element()
+
+    # Save proximal operators
+    prox_dual = [fi.convex_conj.proximal for fi in f]
+    prox_primal = g.proximal
+
+    # run the iterations
+    for k in range(niter):
+
+        # select block
+        selected = fun_select(k)
+
+        # update extrapolation parameter theta
+        theta = float(1 / np.sqrt(1 + 2 * sigma_tilde))
+
+        # update primal variable
+        # tmp = x - tau * z_relax; z_relax used as tmp variable
+        z_relax.lincomb(1, x, -tau, z_relax)
+        # x = prox(tmp)
+        prox_primal(tau)(z_relax, out=x)
+
+        # update dual variable and z, z_relax
+        z_relax.assign(z)
+        for i in selected:
+
+            # compute the step sizes sigma_i based on sigma_tilde
+            sigma_i = sigma_tilde / (
+                    mu[i] * (prob[i] - 2 * (1 - prob[i]) * sigma_tilde))
+
+            # save old yi
+            y_old[i].assign(y[i])
+
+            # tmp = Ai(x)
+            A[i](x, out=y[i])
+
+            # tmp = y_old + sigma_i * Ai(x)
+            y[i].lincomb(1, y_old[i], sigma_i, y[i])
+
+            # hack for the prox operator
+            yy = y[i].copy()
```

Comment:
This issue will be solved with #1201. Maybe make a TODO to remember removing the hack.

Reply:
done, this is now gone.